### DISCUSSION TITLE
how gunicorn + uvicorn + uvloop + httptools

### DISCUSSION THREAD

**Discussion by panla** [Kludex/uvicorn#1143] [Q&A]
```text
i had `pip install gunicorn uvicorn uvloop httptools`, and how to use them all?

in uvicorn origin code

uvicorn.workers.UvicornWorker   :  CONFIG_KWARGS = {"loop": "auto", "http": "auto"}
uvicorn.workers.UvicornH11Worker   :  CONFIG_KWARGS = {"loop": "asyncio", "http": "h11"}
```

- [uvicorn-gunicorn-docker #78](https://github.com/tiangolo/uvicorn-gunicorn-docker/issues/78)

**Comment by euri10** [comment]
Hi @panla 

1.  as you can see in the docs `pip install uvicorn` will install just minimal dependencies (read pure Python deps)
2. you can do what you did to have Cython dependencies, you can also use `pip install uvicorn[standard]` which will install all Cython dependencies.
3. when you use it with gunicorn, by default if you specify nothing (ie you use the default `UvicornWorker` ), we set the loop and the parser on "auto" mode, which in essence is a shortcut to say : if the Cython deps are installed use them, if not use the minimal pure-Python dependencies.
4. If you want another config for the uvicorn worker managed by gunicorn you'll need to subclass it as shown in the docs 

hope that helps

**Reply by panla** [reply]
`worker.py`

```python
class Worker(UvicornWorker):
    CONFIG_KWARGS = {"loop": "uvloop", "http": "httptools"}
```

What I want to know is actually very simple

```text
if `pip install uvicorn httptools uvloop`
    if use `gunicorn -k uvicorn.workers.UvicornWorker`:
        will use httptools and uvloop ?
    elif use `gunicorn -k worker.Worker`:
       will use httptools and uvloop ?
    elif use `uvicorn ....`:
       will use httptools and uvloop ?
```

thanks.

**Reply by maysu1914** [reply]
auto will use uvloop over asyncio and httptools over h11 if they are available

uvicorn/protocols/http/auto.py

uvicorn/loops/auto.py
